src/main/java/com/google/flourbot/entity/EntityModuleImplementation.java

```diff
@@ -1,100 +1,99 @@
 package com.google.flourbot.entity;
 
+import com.google.cloud.firestore.QueryDocumentSnapshot;
 import com.google.flourbot.datastorage.DataStorage;
 import com.google.flourbot.datastorage.FirebaseDataStorage;
-import com.google.cloud.firestore.QueryDocumentSnapshot;
-import java.util.Optional;
-import java.util.Map;
-
 import com.google.flourbot.entity.action.Action;
 import com.google.flourbot.entity.action.SheetAppendAction;
-import com.google.flourbot.entity.trigger.Trigger;
 import com.google.flourbot.entity.trigger.CommandTrigger;
+import com.google.flourbot.entity.trigger.Trigger;
 
-import java.lang.InterruptedException;
+import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 
 public class EntityModuleImplementation implements EntityModule {
-    
-    private static EntityModuleImplementation instance = null;
-    private final DataStorage datastorage;
 
-    private EntityModuleImplementation() {
-        this.datastorage = new FirebaseDataStorage();
+  private static EntityModuleImplementation instance = null;
+  private final DataStorage datastorage;
+
+  private EntityModuleImplementation() {
+    this.datastorage = new FirebaseDataStorage();
+  }
+
+  public Optional<Macro> getMacro(String userEmail, String macroName)
+      throws InterruptedException, ExecutionException {
+    Optional<QueryDocumentSnapshot> optionalDocument =
+        datastorage.getDocument(userEmail, macroName);
+
+    if (!optionalDocument.isPresent()) {
```

Comment:
well done! return empty optional right away and extract the object for future processing right away - nice and clean.